File: packages/silo-client-java/src/main/java/in/org/quicko/silo/client/media/MediaDeleteReport.java
package in.org.quicko.silo.client.media;

import com.fasterxml.jackson.databind.JsonNode;
import in.org.quicko.silo.client.transport.JsonValues;
import java.util.ArrayList;
import java.util.List;

/**
 * What a bulk delete answers, always with a 200: every id's outcome, never a
 * partial result the caller has to infer from a thrown exception.
 */
public record MediaDeleteReport(List<String> deleted, List<MediaDeleteFailure> failed) {

  public MediaDeleteReport {
    deleted = List.copyOf(deleted);
    failed = List.copyOf(failed);
  }

  public static MediaDeleteReport fromWire(JsonNode body) {
    List<MediaDeleteFailure> failures = new ArrayList<>();
    body.path("failed").forEach(failure -> failures.add(MediaDeleteFailure.fromWire(failure)));
    return new MediaDeleteReport(JsonValues.strings(body.get("deleted")), failures);
  }
}

File: packages/silo-client-dart/lib/src/media/media_delete_report.dart
import '../transport/json_values.dart';
import 'media_delete_failure.dart';

/// Every id's outcome in a bulk delete. A partial success is a value, not a throw.
final class MediaDeleteReport {
  const MediaDeleteReport(this.deleted, this.failed);

  factory MediaDeleteReport.fromJson(Map<String, Object?> json) => MediaDeleteReport(
    JsonValues.strings(json['deleted']),
    JsonValues.objects(json['failed']).map(MediaDeleteFailure.fromJson).toList(),
  );

  final List<String> deleted;
  final List<MediaDeleteFailure> failed;
}
